1.97 beta regression: ICE with "`Res::Err` but no error emitted"

This regression was discovered by the 1.97 beta crater run.

<details>
<summary>Error output</summary>

```
[INFO] [stdout] note: no errors encountered even though delayed bugs were created
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
[INFO] [stdout] note: those delayed bugs will now be shown as internal compiler errors
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] 
[INFO] [stdout] error: internal compiler error: `Res::Err` but no error emitted
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout]     |
[INFO] [stdout] note: delayed at /rustc-dev/96f70397ed7ba8baf8d1bb541a1ba8e06e3d6f57/compiler/rustc_hir_typeck/src/expr.rs:588:32
[INFO] [stdout]          0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
[INFO] [stdout]          1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
[INFO] [stdout]          2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
[INFO] [stdout]          3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
[INFO] [stdout]          4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_call
[INFO] [stdout]          6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
[INFO] [stdout]          8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          9: rustc_hir_typeck::check::check_fn
[INFO] [stdout]         10: rustc_hir_typeck::typeck_with_inspect::{closure#0}
[INFO] [stdout]         11: rustc_query_impl::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
[INFO] [stdout]         12: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, false>
[INFO] [stdout]         13: rustc_query_impl::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         14: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
[INFO] [stdout]         15: rustc_hir_analysis::check_crate
[INFO] [stdout]         16: rustc_interface::passes::analysis
[INFO] [stdout]         17: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, false>
[INFO] [stdout]         18: rustc_query_impl::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         19: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
[INFO] [stdout]         20: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
[INFO] [stdout]         21: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
[INFO] [stdout]         22: <std::sys::thread::unix::Thread>::new::thread_start
[INFO] [stdout]         23: <unknown>
[INFO] [stdout]         24: clone
[INFO] [stdout]       
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] note: rustc 1.97.0-beta.1 (96f70397e 2026-05-26) running on x86_64-unknown-linux-gnu
[INFO] [stdout] error: internal compiler error[E0618]: expected function, found `{type error}`
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^---------------
[INFO] [stdout]     |     |
[INFO] [stdout]     |     call expression requires function
[INFO] [stdout]     |
[INFO] [stdout] note: delayed at /rustc-dev/96f70397ed7ba8baf8d1bb541a1ba8e06e3d6f57/compiler/rustc_hir_typeck/src/callee.rs:747:34
[INFO] [stdout]          0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
[INFO] [stdout]          1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
[INFO] [stdout]          2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
[INFO] [stdout]          3: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_call
[INFO] [stdout]          4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
[INFO] [stdout]          6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
[INFO] [stdout]          7: rustc_hir_typeck::check::check_fn
[INFO] [stdout]          8: rustc_hir_typeck::typeck_with_inspect::{closure#0}
[INFO] [stdout]          9: rustc_query_impl::query_impl::typeck_root::invoke_provider_fn::__rust_begin_short_backtrace
[INFO] [stdout]         10: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, false>
[INFO] [stdout]         11: rustc_query_impl::query_impl::typeck_root::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         12: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
[INFO] [stdout]         13: rustc_hir_analysis::check_crate
[INFO] [stdout]         14: rustc_interface::passes::analysis
[INFO] [stdout]         15: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, false>
[INFO] [stdout]         16: rustc_query_impl::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
[INFO] [stdout]         17: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
[INFO] [stdout]         18: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
[INFO] [stdout]         19: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
[INFO] [stdout]         20: <std::sys::thread::unix::Thread>::new::thread_start
[INFO] [stdout]         21: <unknown>
[INFO] [stdout]         22: clone
[INFO] [stdout]       
[INFO] [stdout]    --> src/sync_api.rs:403:5
[INFO] [stdout]     |
[INFO] [stdout] 403 |     crate::count_tokens( text, model )
[INFO] [stdout]     |     ^^^^^^^^^^^^^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stderr] 
[INFO] [stderr] note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2
[INFO] [stderr] 
[INFO] [stderr] note: some of the compiler flags provided by cargo are hidden
[INFO] [stderr] 
[INFO] [stderr] query stack during panic:
[INFO] [stderr] end of query stack
[INFO] [stderr] error: could not compile `api_xai`
```

</details>

https://crater-reports.s3.amazonaws.com/beta-1.97-3-v2/1.97.0-beta.1/reg/api_xai-0.3.0/log.txt

Minimized: [repro.zip](https://github.com/user-attachments/files/28657953/repro.zip)

src/lib.rs
```rust
use dep::my_macro;

pub fn foo() {
    ambiguous();
}

mod submodule {
    pub fn ambiguous() {}
}

pub mod ambiguous {}

my_macro!();
```

dep/src/lib.rs
```rust
#[proc_macro]
pub fn my_macro(_: proc_macro::TokenStream) -> proc_macro::TokenStream {
    r"
        use own::*;
        mod own {
            pub use super::submodule::*;
            pub use super::ambiguous;
        }
    "
    .parse()
    .unwrap()
}
```

<details>
<summary>Error output on beta</summary>

```
note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: `Res::Err` but no error emitted
 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^
  |
note: delayed at /rustc-dev/0417c25868d6dfbd1c291dfeae950504faa6f790/compiler/rustc_hir_typeck/src/expr.rs:588:32
         0: <std::backtrace::Backtrace>::create
         1: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         2: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         3: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         4: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, &str>
         5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_path
         6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
         8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
        10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
        11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_or_body_tail
        12: rustc_hir_typeck::check::check_fn
        13: rustc_hir_typeck::typeck_with_inspect::{closure#0}
        14: rustc_hir_typeck::typeck_root
        15: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, true>
        16: rustc_query_impl::query_impl::typeck_root::execute_query_incr::__rust_end_short_backtrace
        17: <rustc_middle::query::plumbing::TyCtxtEnsureOk>::typeck::<rustc_span::def_id::LocalDefId>
        18: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
        19: rustc_hir_analysis::check_crate
        20: rustc_interface::passes::analysis
        21: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, true>
        22: rustc_query_impl::query_impl::analysis::execute_query_incr::__rust_end_short_backtrace
        23: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
        24: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        25: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        26: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        27: <std::sys::thread::unix::Thread>::new::thread_start
        28: __pthread_cond_wait

 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^

error: internal compiler error[E0618]: expected function, found `{type error}`
 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^--
  |     |
  |     call expression requires function
  |
note: delayed at /rustc-dev/0417c25868d6dfbd1c291dfeae950504faa6f790/compiler/rustc_hir_typeck/src/callee.rs:747:34
         0: <std::backtrace::Backtrace>::create
         1: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         2: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         3: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::report_invalid_callee
         5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
         6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_block
         8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
         9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_or_body_tail
        10: rustc_hir_typeck::check::check_fn
        11: rustc_hir_typeck::typeck_with_inspect::{closure#0}
        12: rustc_hir_typeck::typeck_root
        13: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 8]>, rustc_middle::dep_graph::graph::DepNodeIndex>, true>
        14: rustc_query_impl::query_impl::typeck_root::execute_query_incr::__rust_end_short_backtrace
        15: <rustc_middle::query::plumbing::TyCtxtEnsureOk>::typeck::<rustc_span::def_id::LocalDefId>
        16: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis::check_crate::{closure#1}>::{closure#0}
        17: rustc_hir_analysis::check_crate
        18: rustc_interface::passes::analysis
        19: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, true>
        20: rustc_query_impl::query_impl::analysis::execute_query_incr::__rust_end_short_backtrace
        21: rustc_interface::passes::create_and_enter_global_ctxt::<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>
        22: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        23: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        24: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        25: <std::sys::thread::unix::Thread>::new::thread_start
        26: __pthread_cond_wait

 --> src/lib.rs:4:5
  |
4 |     ambiguous();
  |     ^^^^^^^^^

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.97.0-beta.2 (0417c2586 2026-05-30) running on aarch64-apple-darwin

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `repro` (lib)
```

</details>

Bisection gave the same strange results as before.

--- a/compiler/rustc_resolve/src/imports.rs
+++ b/compiler/rustc_resolve/src/imports.rs
@@ -628,6 +628,15 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             module.underscore_disambiguator.get()
         });
         self.update_local_resolution(module, key, orig_ident_span, |this, resolution| {
+            if res == Res::Err
+                && let Some(old_decl) = resolution.best_decl()
+                && old_decl.res() != Res::Err
+            {
+                // Do not override real declarations with `Res::Err`s from error recovery.
+                // FIXME: this special case shouldn't be necessary, but removing it triggers an ICE
+                // due to some other issues (#157406, tests/ui/imports/dummy-import-ice.rs).
+                return Ok(());
+            }
             if decl.is_glob_import() {
                 resolution.glob_decl = Some(match resolution.glob_decl {
                     Some(old_decl) => this.select_glob_decl(old_decl, decl),

--- a/tests/ui/imports/auxiliary/dummy-import-ice-macro.rs
+++ b/tests/ui/imports/auxiliary/dummy-import-ice-macro.rs
@@ -1,0 +1,15 @@
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+#[proc_macro]
+pub fn my_macro(_: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    r"
+        use own::*;
+        mod own {
+            pub use super::submodule::*;
+            pub use super::ambiguous;
+        }
+    "
+    .parse()
+    .unwrap()
+}

--- a/tests/ui/imports/dummy-import-ice.rs
+++ b/tests/ui/imports/dummy-import-ice.rs
@@ -1,0 +1,20 @@
+// Regression test for issue #157406.
+
+//@ check-pass
+//@ proc-macro: dummy-import-ice-macro.rs
+
+extern crate dummy_import_ice_macro;
+
+pub fn foo() {
+    ambiguous();
+}
+
+mod submodule {
+    pub fn ambiguous() {}
+}
+
+pub mod ambiguous {}
+
+dummy_import_ice_macro::my_macro!();
+
+fn main() {}

--- a/tests/ui/imports/issue-56125.rs
+++ b/tests/ui/imports/issue-56125.rs
@@ -15,7 +15,7 @@ mod m2 {
 mod m3 {
     mod empty {}
     use empty::issue_56125; //~ ERROR unresolved import `empty::issue_56125`
-    use issue_56125::*;
+    use issue_56125::*; //~ ERROR `issue_56125` is ambiguous
 }
 
 fn main() {}

--- a/tests/ui/imports/issue-56125.stderr
+++ b/tests/ui/imports/issue-56125.stderr
@@ -56,7 +56,24 @@ LL |     use issue_56125::non_last_segment::non_last_segment::*;
    = help: consider adding an explicit import of `issue_56125` to disambiguate
    = help: or use `self::issue_56125` to refer to this module unambiguously
 
-error: aborting due to 3 previous errors
+error[E0659]: `issue_56125` is ambiguous
+  --> $DIR/issue-56125.rs:18:9
+   |
+LL |     use issue_56125::*;
+   |         ^^^^^^^^^^^ ambiguous name
+   |
+   = note: ambiguous because of a conflict between a name from a glob import and an outer scope during import or macro resolution
+   = note: `issue_56125` could refer to a crate passed with `--extern`
+   = help: use `::issue_56125` to refer to this crate unambiguously
+note: `issue_56125` could also refer to the module imported here
+  --> $DIR/issue-56125.rs:18:9
+   |
+LL |     use issue_56125::*;
+   |         ^^^^^^^^^^^^^^
+   = help: consider adding an explicit import of `issue_56125` to disambiguate
+   = help: or use `self::issue_56125` to refer to this module unambiguously
+
+error: aborting due to 4 previous errors
 
 Some errors have detailed explanations: E0432, E0659.
 For more information about an error, try `rustc --explain E0432`.

--- a/tests/ui/imports/shadow-glob-module-resolution-2.rs
+++ b/tests/ui/imports/shadow-glob-module-resolution-2.rs
@@ -14,5 +14,7 @@ use a::*;
 use e as b;
 //~^ ERROR: unresolved import `e`
 use b::c::D as e;
+//~^ ERROR: cannot determine resolution for the import
+//~| ERROR: cannot determine resolution for the import
 
 fn main() { }

--- a/tests/ui/imports/shadow-glob-module-resolution-2.stderr
+++ b/tests/ui/imports/shadow-glob-module-resolution-2.stderr
@@ -1,3 +1,17 @@
+error: cannot determine resolution for the import
+  --> $DIR/shadow-glob-module-resolution-2.rs:16:5
+   |
+LL | use b::c::D as e;
+   |     ^^^^^^^^^^^^
+
+error: cannot determine resolution for the import
+  --> $DIR/shadow-glob-module-resolution-2.rs:16:5
+   |
+LL | use b::c::D as e;
+   |     ^^^^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
 error[E0432]: unresolved import `e`
   --> $DIR/shadow-glob-module-resolution-2.rs:14:5
    |
@@ -12,6 +26,6 @@ LL - use e as b;
 LL + use a as b;
    |
 
-error: aborting due to 1 previous error
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0432`.

--- a/tests/ui/imports/shadow-glob-module-resolution-4.rs
+++ b/tests/ui/imports/shadow-glob-module-resolution-4.rs
@@ -12,6 +12,8 @@ use e as b;
 
 use b::C as e;
 //~^ ERROR: unresolved import `b::C`
+//~| ERROR: cannot determine resolution for the import
+//~| ERROR: cannot determine resolution for the import
 
 fn e() {}
 

--- a/tests/ui/imports/shadow-glob-module-resolution-4.stderr
+++ b/tests/ui/imports/shadow-glob-module-resolution-4.stderr
@@ -1,9 +1,23 @@
+error: cannot determine resolution for the import
+  --> $DIR/shadow-glob-module-resolution-4.rs:13:5
+   |
+LL | use b::C as e;
+   |     ^^^^^^^^^
+
+error: cannot determine resolution for the import
+  --> $DIR/shadow-glob-module-resolution-4.rs:13:5
+   |
+LL | use b::C as e;
+   |     ^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
 error[E0432]: unresolved import `b::C`
   --> $DIR/shadow-glob-module-resolution-4.rs:13:5
    |
 LL | use b::C as e;
    |     ^^^^^^^^^
 
-error: aborting due to 1 previous error
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0432`.

